libbeat/processors/add_process_metadata/add_process_metadata.go

```diff
@@ -93,18 +101,44 @@ func newProcessMetadataProcessorWithProvider(cfg *common.Config, provider proces
 		return nil, errors.Wrapf(err, "fail to unpack the %v configuration", processorName)
 	}
 
-	p := addProcessMetadata{
-		config:   config,
-		provider: provider,
-		log:      log,
-	}
-	if p.mappings, err = config.getMappings(); err != nil {
+	mappings, err := config.getMappings()
+
+	if err != nil {
 		return nil, errors.Wrapf(err, "error unpacking %v.target_fields", processorName)
 	}
 
+	var cidProvider gosigarCidProvider
+
+	// don't use cgroup.ProcessCgroupPaths to save it from doing the work when container id disabled
+	if ok := containsValue(mappings, "container.id"); ok {
+		cidProvider = newCidProvider(config.HostPath, config.CgroupPrefixes, processCgroupPaths)
+	} else {
```

Comment:
> The reason for that is I felt that the use of nil is a code smell, as we could not extend nil in the future. Also, the extra steps introduced by the dummy function will probably be optimized away by the compiler.

I still think that leaving this `cidProvider` as nil is simpler when the feature is disabled. It doesn't leave anything to chance with compiler optimizations.

I'm not sure what you meant by "as we could not extend nil in the future".

Reply:
Just a small style different, I changed it to use nil.

Trying to explain "as we could not extend nil in the future". 

* Put the `cidProvider` to nil, in the case of `container.id` is disabled,  Then there has to be a nil check in the `enrich` function. If there is another case, there has to be another type of check for what `cidProvider`.
* Whereas, by changing the behavior of the `cidProvider` so that when `container.id` is disabled it will not add container id. If there is another case, just pass an instance of `cidProvider` which behaved differently, then there is not any change required in the `enrich`.
* Furthermore, say if there is any change required in the runtime. It would be possible to pass a different instance of `cidProvider` to the processor determined by some conditions.

It is not specifically required in this PR, but just thought it could leave more flexibility

Reply:
why not just overwrite/expand existing provider instead of creating another one? Some functionalities might still go for regular provider and e.g. try to read `/proc/$some_pid/exe` instead `$host_path/proc/$some_pid/exe`